### PR TITLE
Welcome vaccine data

### PR DESCRIPTION
**Pull request by Jimmy-Tempest**
image

**Features**
- Vaccination statistic for the welcome page

### CODE REVIEW

**Review by Temp9699 — CHANGES_REQUESTED** [review]
From the [Preview](https://mysk-frontend-git-feature-welcome-cb68fa-suankularb-developers.vercel.app/)

The "Add" button doesn't seem to do anything
The "Novavax" option for the Provider part got cut off at the bottom
The labels also appears as the picture attached
Screenshot 2022-11-19 233948

**Review by Temp9699 — APPROVED** [review]
LGTM

File: src/pages/welcome.tsx [hunk 1]
```diff
@@ -471,6 +483,353 @@ const DataCheckSection = ({
...
+      ref={sectionRef}
+    >
+      <Section>
+        <Header
+          icon={<MaterialIcon icon="badge" allowCustomSize />}
+          text={t("welcome.vaccineData.title")}
+        />
+        <LayoutGridCols cols={6}>
+          <div className="col-span-3">
+            <Card type="stacked" appearance="outlined">
```
Comment (SiravitPhokeed) [inline src/pages/welcome.tsx:559]: Currently Dropdown Options overflow this card, and the lowest option is not visible. Try adding `className="overflow-visible"`.

File: src/pages/welcome.tsx [hunk 2]
```diff
@@ -471,6 +483,353 @@ const DataCheckSection = ({
...
+                  }}
+                  disabled={disabled || !validateForm() || loading}
+                />
+              </CardActions>
+            </Card>
+          </div>
+
+          <div className="col-span-3 md:col-start-5">
+            {vaccineData.length > 0 &&
+              vaccineData.map((vaccine) => (
```
Comment (SiravitPhokeed) [inline src/pages/welcome.tsx:655]: Personally I would take a few minutes out of my day to do some animations for this. You know why I said few minutes? You can literally copy the code from Create Form. It’s almost identical.

File: src/pages/welcome.tsx [hunk 3]
```diff
@@ -471,6 +483,353 @@ const DataCheckSection = ({
...
+                            label: "Vaxzevria (AstraZeneca)",
+                          },
+                          { value: "spikevax", label: "Spikevax (Moderna)" },
+                          { value: "jcovden", label: "Jcovden (J&J)" },
+                          { value: "covilo", label: "Covilo (Sinopharm)" },
+                          {
+                            value: "covovax",
+                            label: "COVOVAX (Novavax formulation)",
+                          },
+                        ]}
```
Comment (SiravitPhokeed) [inline src/pages/welcome.tsx:734]: You’ve used this twice now. Mayhaps a constant?

File: src/utils/backend/vaccine.ts [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+  personId: number
+): Promise<VaccineRecord[]> {
+  const { data, error } = await supabase
+    .from("vaccine_records")
+    .select("*")
+    .eq("person", personId)
+    .order("vaccination_date", { ascending: true });
+
+  if (error) {
+    throw error;
```
Comment (SiravitPhokeed) [inline src/utils/backend/vaccine.ts:15]: If you wanna try something new, you should probably ask first. I don’t know anything about this way of reporting errors. If you think this is how we should be moving forward, let’s talk, and then we’ll see how we can migrate.

File: src/utils/backend/vaccine.ts [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+    data!.map((vaccineRecord, index) => ({
+      id: vaccineRecord.id,
+      doseNo: index + 1,
+      vaccineName: vaccineRecord.vaccine_name,
+      vaccineDate: new Date(vaccineRecord.vaccination_date),
+      lotNo: vaccineRecord.lot_no,
+      administeredBy: vaccineRecord.administering_center,
+    })) ?? []
+  );
+}
```
Comment (SiravitPhokeed) [inline src/utils/backend/vaccine.ts:28]: Now that you brought it up, holy bananas 🍌🍌 is this a better way to do this. Death to `database.ts`!

File: src/pages/welcome.tsx [hunk 6]
```diff
@@ -471,6 +483,353 @@ const DataCheckSection = ({
...
+                      },
+                      personid!
+                    );
+                    const newVaccineRecords = await getVaccineRecordbyPersonId(
+                      supabase,
+                      personid!
+                    );
+                    toggleLoading();
+                    setVaccineData(newVaccineRecords);
+                  }}
```
Comment (SiravitPhokeed) [inline src/pages/welcome.tsx:646]: Don’t you agree that altering the database here makes things much more complicate (for us) and slow (for users)? Why not just keep track of everything in states and just do all the changes afterwards?

File: src/pages/welcome.tsx [hunk 7]
```diff
@@ -471,6 +483,353 @@ const DataCheckSection = ({
...
+                  disabled={disabled || !validateForm() || loading}
+                />
+              </CardActions>
+            </Card>
+          </div>
+
+          <div className="col-span-3 md:col-start-5">
+            {vaccineData.length > 0 &&
```
Comment (SiravitPhokeed) [inline src/pages/welcome.tsx:654]: So nothing appears when there are no records? I think this Section would feel more complete if there’s a box on the right saying there are no records. Too much white space in a design is not a good thing.

File: src/pages/welcome.tsx [hunk 8]
```diff
@@ -471,6 +483,353 @@ const DataCheckSection = ({
...
+              vaccineData.map((vaccine) => (
+                <Card
+                  type="stacked"
+                  appearance="tonal"
+                  className="mb-4"
```
Comment (SiravitPhokeed) [inline src/pages/welcome.tsx:659]: Why not just use `gap` in the parent element?

**Review by SiravitPhokeed — CHANGES_REQUESTED** [review]
Quite a few things. Big things, small things. I know you said you just want to get this done, but I don’t have it in me to let this ship.

Overall: this might be more complicated than it had to be, but I can’t be 100% sure because you set up the database stuff.

One more thing (that I can’t comment where I want to), Up Next Cards! This is a new Section, so it should be reflected there as well. We’ll figure out what to do with the whole Up Next Section another time.

File: src/pages/welcome.tsx (hunk 6 above)
Comment (Jimmy-Tempest) [reply src/pages/welcome.tsx:646]: If you ask me as a user, I would expect when I click add the data should reflect on the database the same way as it does on the screen. For the complicated part, I want to make sure any data being display is accurate to the backend so I just call the backend to get the data. If speed really become an issue then we could optout manually add the new record to the state then resort everything to get the actual dose number.

Comment (SiravitPhokeed) [reply src/pages/welcome.tsx:646]: But it literally says “Save and continue,” though, implying that your changes are not saved yet. The fact that the changes are saved confused me (I noticed from the delay in deleting) so I thought to check the code.
I still don’t understand how this is a better approach for anyone. Working with the backend sucks and I’m pretty sure you’re supposed to contact the backend as few times as possible. Anyway, I know you want to get this out fast so we’ll just leave this for the future.

File: public/static/locales/th/landing.json [hunk 9]
```diff
@@ -55,6 +55,19 @@
...
+        "vaccineCard": "วัคซีนเข็มที่ {{doseNo}}"
+      },
+      "action": {
+        "continue": "บันทึกและไปต่อ",
+        "add": "เพิ่ม"
+      }
+    },
     "preparingForStudents": {
       "title": "การเตรียมระบบให้นักเรียน",
       "notice": "สำคัญ: ให้อาจารย์อ่านส่วนนี้ก่อนเข้าใช้งาน",
```
[the remark this answers is not in the document]
Comment (Jimmy-Tempest) [reply public/static/locales/th/landing.json:73]: the whole `preparingForStudents` part?

File: src/pages/welcome.tsx [hunk 10]
```diff
@@ -471,6 +483,353 @@ const DataCheckSection = ({
...
+                    toggleLoading();
+                    setVaccineData(newVaccineRecords);
+                  }}
+                  disabled={disabled || !validateForm() || loading}
+                />
+              </CardActions>
+            </Card>
+          </div>
+
+          <div className="col-span-3 md:col-start-5">
```
[the remark this answers is not in the document]
Comment (Jimmy-Tempest) [reply src/pages/welcome.tsx:653]: I am not sure if I understand the implementation of this "fill the column" thing

File: public/static/locales/th/landing.json (hunk 9 above)
[the remark this answers is not in the document]
Comment (SiravitPhokeed) [reply public/static/locales/th/landing.json:73]: Yes.

File: src/pages/welcome.tsx (hunk 10 above)
[the remark this answers is not in the document]
Comment (SiravitPhokeed) [reply src/pages/welcome.tsx:653]: Replace that line with this:
```html
<div>
```

There was no point on spanning this section into many columns then starting on the fifth column. We're just making a 2 column grid. Don't make it complicated.

File: src/utils/backend/vaccine.ts [hunk 11]
```diff
@@ -0,0 +1,86 @@
...
+): Promise<VaccineRecord[]> {
+  const { data, error } = await supabase
+    .from("vaccine_records")
+    .select("*")
+    .eq("person", personId)
+    .order("vaccination_date", { ascending: true });
+
+  if (error) {
+    // throw error;
+    console.error(error);
```
Comment (SiravitPhokeed) [inline src/utils/backend/vaccine.ts:16]: This is not what I meant. We’re using `BackendReturn` and `BackendDataReturn` for most backend functions right now. This ensures all backend functions have predictable returns.

None of the functions here use it correctly. I’ll let it be your call to fix this in this PR or not.

File: src/utils/backend/vaccine.ts [hunk 12]
```diff
@@ -0,0 +1,86 @@
...
+    vaccination_date: vaccineRecord.vaccineDate,
+    lot_no: vaccineRecord.lotNo,
+    administering_center: vaccineRecord.administeredBy,
+  }));
+
+  const { data, error } = await supabase
+    .from("vaccine_records")
+    .upsert(dbVaccineRecords, { onConflict: "id" });
+
+  return { data: [], error };
```
Comment (SiravitPhokeed) [inline src/utils/backend/vaccine.ts:85]: If there’s always nothing to return, use `BackendReturn`.

**Review by SiravitPhokeed — APPROVED** [review]
_LGTM!_ All further changes don’t directly impact the user so I’ll allow the merge.

**Review by Temp9699 — COMMENTED** [review]
LGTM

File: src/utils/backend/vaccine.ts [hunk 13]
```diff
@@ -0,0 +1,86 @@
...
+): Promise<BackendDataReturn<VaccineRecord[]>> {
+  const { data, error } = await supabase
+    .from("vaccine_records")
+    .select("*")
+    .eq("person", personId)
+    .order("vaccination_date", { ascending: true });
+
+  if (error) {
+    // throw error;
+    // console.error(error);
```
Comment (SiravitPhokeed) [inline src/utils/backend/vaccine.ts:20]: We usually log the errors too when they happen, so this line was actually correct.

**Review by SiravitPhokeed — APPROVED** [review]
_LGTM!_ For the same reason.

**Pull request merged**